docs/blueprints/001-installation-of-extensions.md

```diff
@@ -0,0 +1,114 @@
+
+The installer should make a clear distinction between basic mandatory components and extensions for 3rd party tools, but at the same time continue to provide the "all-in-one" experience of installing everything in one shot. This is a summary of needed installer changes:
+
+* installer command line arguments that the user can supply to indicate which optional extensions need to be installed (e.g. fuseml-install --extension mlflow --extension kfserving)
+* make it possible to install/uninstall extensions in follow-up installer invocations, separately from the main installation
+* MLFlow no longer installed by default
+* add support to install KFServing and KNative.
+
+## Current state
+
+`fuseml-installer` has two ways of installing components
+
+* `fuseml-installer install` - installs all necessary components (but actually _also_ MLFlow, which should be an extension)
+* `make` targets. Specific targets defined in the `Makefile` describe the installation of other components, like KFServing
+
+## Proposed solution
+
+### Makefile
+
+Get rid of specific installation targets that install something for production usage (e.g. not used by testing/development).
+
+### CLI
+
+1. Implement new option for `install` command
+
+  `fuseml-installer install [--extension <name>] [--extension-repository <path>]`
+
+  This command installs all necessary components AND extensions selected by `--extension` option. Optionally uses path specified
+  by `--extension-repository` option as a source for extensions that should be installed. Path could be local directory on URL.
+
+
+2. Implement new command for managing extensions. This could be useful for adding an extension to already configured system.
+
+  `fuseml-installer extension`
+
+  This should have several sub commands, like
+
+  * `fuseml-installer extension list` - list installed extensions
+  * `fuseml-installer extension available [--extension-repository <path>]` = list extensions available for installation
+  * `fuseml-installer extension add <name> [--extension-repository <path>]` - install selected extension
+  * `fuseml-installer extension remove <name> [--extension-repository <path>]` - ubinstalls selected extension
+
+
+### Description files
+
+Extension repository is a directory (local or remote) containing description files of extensions. Each description file provides 
+description of operations that could be done with the extension.
+
+Each extension has its own subdirectory that matches the extension name. Mandatory component under each extension subdirectory is the
+description file, in `yaml` format.
+
+Directory structure:
+
+```
+extensions
+ - installation
+    - mlflow
+      - description.yaml
+    - knative
+      - description.yaml
+      - install.sh
+      - uninstall.sh
+```
+
+Possible extension types are
```

Comment:
I wouldn't call this an extension type, but the type of installation for an extension. You should reserve the name extension type to be used for things like prediction platforms, model stores, etc. But not my point here...

My advice would be to try and see if this maps well to the use-cases we already have: KNative, KFServing and maybe even Seldon Core. I think often you need to combine several installation types and run them in a sort of check-list manner: first, you install a helm chart, then you create some resources via manifests, then maybe you run some command or script. So instead having one installation type here, maybe it would be more flexible to have an ordered list of installation steps.

Of course, you might say that anything that's more complex than a helm chart or a single k8s manifest should be implemented as a shell script.